File: agent/control/exceptions.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass
class TurnPaused(Exception):
    interaction: dict[str, Any]
    tool_messages: list[dict[str, Any]] = field(default_factory=list)

    def __str__(self) -> str:
        kind = self.interaction.get("kind") or "interaction"
        ident = self.interaction.get("id") or "unknown"
        return f"turn paused for {kind}: {ident}"

File: packages/core/src/control/exceptions.ts
/**
 * 控制流异常 (MIG-CTRL-011)。对齐 Python `agent/control/exceptions.py`。
 * TurnPaused: 暂停当前回合，携带 interaction + tool_messages。
 */
export class TurnPaused extends Error {
  readonly interaction: Record<string, unknown>
  readonly toolMessages: Array<Record<string, unknown>>

  constructor(
    interaction: Record<string, unknown>,
    toolMessages: Array<Record<string, unknown>> = [],
  ) {
    const kind = interaction.kind ?? 'interaction'
    const ident = interaction.id ?? 'unknown'
    super(`turn paused for ${kind}: ${ident}`)
    this.name = 'TurnPaused'
    this.interaction = interaction
    this.toolMessages = toolMessages
  }
}
